pool/src/graph.rs

```diff
@@ -179,6 +180,31 @@ impl DirectedGraph {
 		}
 	}
 
+	/// Promote any non-root vertices to roots based on current edges.
+	/// For a given tx, if there are no edges with that tx as destination then it is a root.
+	pub fn update_roots(&mut self) {
+		let mut new_roots = vec![];
+		for x in &self.vertices {
+			let edges = &self.edges.values().collect::<Vec<_>>();
```

Comment:
I think you can move this outside the for loop to avoid allocating for every vertex.